<a href="https://colab.research.google.com/github/helvetia1975/JAY_INFO4670_SPRING2026_ASSIGNMENT2/blob/main/Assignment_2_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#1. Data Quality Check

1.1 Using Python (pandas, matplotlib, or seaborn), load and inspect the Assignment 2 dataset.

In [129]:
# Write your code here

from google.colab import files
uploaded = files.upload()


Saving Assignment 2 dataset.csv to Assignment 2 dataset (3).csv


Write code to explore the data distribution (e.g., region, type, year) and check whether there is any bias. Provide both the code and your interpretation.

In [130]:
# Write your code here

import pandas as pd
import numpy as np

df = pd.read_csv('Assignment 2 dataset.csv')

region_counts = df['region'].value_counts() # count how many times a region appears
print("\nRegion Distribution:")
print(region_counts)

print("\nRegion Statistics:") # calculates mean and standard deviation of region
print("Mean:", region_counts.mean())
print("Standard Deviation:", region_counts.std())

type_counts = df['type'].value_counts() # count how many times a type appears
type_percent = df['type'].value_counts(normalize=True) * 100 # type count as a percentage

print("\nType Distribution:")
print(type_counts)

print("\nType Percentage:")
print(type_percent)

year_counts = df['year'].value_counts().sort_index() # counts how many times a year appears
print("\nYear Distribution:")
print(year_counts)


Region Distribution:
region
WestTexNewMexico       340
Albany                 338
BaltimoreWashington    338
Boise                  338
Boston                 338
Atlanta                338
California             338
Charlotte              338
Chicago                338
CincinnatiDayton       338
Columbus               338
DallasFtWorth          338
Denver                 338
Detroit                338
GrandRapids            338
GreatLakes             338
HarrisburgScranton     338
HartfordSpringfield    338
Houston                338
Indianapolis           338
Jacksonville           338
BuffaloRochester       338
LasVegas               338
LosAngeles             338
MiamiFtLauderdale      338
Louisville             338
Nashville              338
NewOrleansMobile       338
NewYork                338
Midsouth               338
NorthernNewEngland     338
Orlando                338
Philadelphia           338
PhoenixTucson          338
Pittsburgh             338
Plains                 338

My code displays the frequencies for each entry in the region, type, and year columns.
For the region column, we see that each region in the dataset appears mostly the same amount, with the exception of WestTexNewMexico at 340 times. It is fair to say that there is a minimal amount of bias in the region column.
For the type column, we see that there are only two types, either organic or conventional. Their count is very close to each other, meaning data is distributed almost evenly.
For the year column, we see some variation among the listed years. Noticeably, we see 1904 and 2018 have considerably less appearances than the others. This indicates there is bias in this column.

1.2 Write Python code to check for duplicate rows and missing values in the dataset. Show the number of duplicates and missing values for each column. Then, explain (in comments or markdown) how you would handle these issues (e.g., drop, impute, or replace).

In [131]:
# Write your code here

import pandas as pd
import numpy as np

df = pd.read_csv('Assignment 2 dataset.csv')

print("\nDuplicate Rows:")
num_duplicates = df.duplicated().sum() # check for duplicate rows
if num_duplicates > 0: # display the duplicatae rows
    print(df[df.duplicated(keep=False)]) # checks for duplicate row (if duplicate, mark as True) and only displays rows that return as True
print(f"\nNumber of duplicates: {num_duplicates}")

print("\nMissing Values by Column:") # check for missing entries
missing_values = df.isnull().sum()
print(missing_values)

print(f"\nTotal missing values: {missing_values.sum()}")


Duplicate Rows:
       Column 1       Date  AveragePrice  Total Volume     4046     4225  \
18244         7   2-4-2018          1.63      17074.83  2046.96  1529.20   
18247        10  1-14-2018          1.93      16205.22  1527.63  2981.04   
18249        10  1-14-2018          1.93      16205.22  1527.63  2981.04   
18253         7   2-4-2018          1.63      17074.83  2046.96  1529.20   

         4770  Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
18244    0.00    13498.67    13066.82      431.85          0.0  organic  2018   
18247  727.01    10969.54    10919.54       50.00          0.0  organic  2018   
18249  727.01    10969.54    10919.54       50.00          0.0  organic  2018   
18253    0.00    13498.67    13066.82      431.85          0.0  organic  2018   

                 region  
18244  WestTexNewMexico  
18247  WestTexNewMexico  
18249  WestTexNewMexico  
18253  WestTexNewMexico  

Number of duplicates: 2

Missing Values by Column:
Column 1      

For the duplicate rows, I would simply drop any duplicate rows. As for the missing values, you could drop the row if it has missing data, but you would lose data. Instead of dropping the data, you could instead insert the mean of the data or the mode of the data.

1.3 Use Python code to print the number of rows and columns in the dataset (e.g., with df.shape). Based on the dataset size, explain (briefly) whether you think the dataset is sufficient for training a machine learning model.

In [132]:
# Write your code here

import pandas as pd
import numpy as np

df = pd.read_csv('Assignment 2 dataset.csv')

print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

Number of rows: 18254
Number of columns: 14


This dataset is suitable for a machine learning model, since it has over 1000 samples and has more than 10 features (columns). This lets the model learn with sufficient amount of information. There are also minimal amounts missing data, bias, and duplicates so data quality is good.

#2. Data Cleaning and Preprocessing

2.1 Remove the first column or “Column 1” from the dataset. Treat the ‘year’ variable as nominal.

In [133]:
# Write your code here

import pandas as pd
import numpy as np

df = pd.read_csv('Assignment 2 dataset.csv')

print("Before cleaning:") # dataset before changes
print(f"Shape: {df.shape}")
print(f"Columns: {list(df.columns)}")

df = df.drop(df.columns[0], axis=1) # remove the first column "Column 1"

df['year'] = df['year'].astype(str) # change year to nominal

print("\nAfter cleaning:") # dataset after changes
print(f"Shape: {df.shape}")
print(f"Year dtype: {df['year'].dtype}") # check year type
print(f"Columns: {list(df.columns)}")

Before cleaning:
Shape: (18254, 14)
Columns: ['Column 1', 'Date', 'AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags', 'Small Bags', 'Large Bags', 'XLarge Bags', 'type', 'year', 'region']

After cleaning:
Shape: (18254, 13)
Year dtype: object
Columns: ['Date', 'AveragePrice', 'Total Volume', '4046', '4225', '4770', 'Total Bags', 'Small Bags', 'Large Bags', 'XLarge Bags', 'type', 'year', 'region']


2.2 Check for duplicate values and remove them.

In [134]:
# Write your code here

import pandas as pd
import numpy as np

print("Before removing duplicates:") # dataset before potential drops
print(f"Total rows: {len(df)}")
duplicates = df.duplicated().sum() # check for duplicates
if num_duplicates > 0: # display the duplicatae rows
    print(df[df.duplicated(keep=False)]) # checks for duplicate row (if duplicate, mark as True) and only displays rows that return as True
print(f"Duplicate rows: {duplicates}") # list dupliactes

df = df.drop_duplicates() # drop duplicates

print("\nAfter removing duplicates:") # check if duplicated rows were deleted
print(f"Total rows: {len(df)}")
print(f"Rows removed: {duplicates}")

Before removing duplicates:
Total rows: 18254
            Date  AveragePrice  Total Volume     4046     4225    4770  \
18244   2-4-2018          1.63      17074.83  2046.96  1529.20    0.00   
18247  1-14-2018          1.93      16205.22  1527.63  2981.04  727.01   
18249  1-14-2018          1.93      16205.22  1527.63  2981.04  727.01   
18253   2-4-2018          1.63      17074.83  2046.96  1529.20    0.00   

       Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
18244    13498.67    13066.82      431.85          0.0  organic  2018   
18247    10969.54    10919.54       50.00          0.0  organic  2018   
18249    10969.54    10919.54       50.00          0.0  organic  2018   
18253    13498.67    13066.82      431.85          0.0  organic  2018   

                 region  
18244  WestTexNewMexico  
18247  WestTexNewMexico  
18249  WestTexNewMexico  
18253  WestTexNewMexico  
Duplicate rows: 2

After removing duplicates:
Total rows: 18252
Rows removed: 2


2.3 Check for missing values. If a data record (row) only has a few missing values, replace the missing values with the median of the column feature in that specific “Region” variable. If most column values in a data record are missing, remove the data record.

In [135]:
# Write your code here

import pandas as pd
import numpy as np

print("Initial missing values:") # initial dataset state
print(df.isnull().sum())
print(f"\nTotal missing: {df.isnull().sum().sum()}")
print(f"Initial dataset size: {len(df)} rows")

missing_per_row = df.isnull().sum(axis=1) # remove rows with more than 50% of values missing
threshold = len(df.columns) * 0.5  # 50% threshold
rows_to_remove = missing_per_row > threshold

print(f"\nRows with greater than {threshold:.0f} missing values: {rows_to_remove.sum()}")
df = df[~rows_to_remove] # the NOT operator (~) flips Boolean values (False -> True and True -> False); this makes the rows_to_remove rows False and df[] keeps only True rows

numeric_cols = df.select_dtypes(include=[np.number]).columns # select only numeric columns

for col in numeric_cols: # replace missing values with a region-specific median
    if df[col].isnull().any():
        for region in df['region'].unique(): # loop through the region column and list each unique region
            region_median = df[df['region'] == region][col].median() # calculate the median for that region

            mask = (df['region'] == region) & (df[col].isnull()) # checks if the row is in the same region and is missing a value
            df.loc[mask, col] = region_median # selects specific cells with missing values and have the same region, and fills that cell with the median for that region

print("\nFinal missing values:")
print(df.isnull().sum())
print(f"\nTotal missing: {df.isnull().sum().sum()}")
print(f"Final dataset size: {len(df)} rows")

Initial missing values:
Date            0
AveragePrice    0
Total Volume    1
4046            2
4225            1
4770            1
Total Bags      1
Small Bags      2
Large Bags      2
XLarge Bags     1
type            1
year            0
region          0
dtype: int64

Total missing: 12
Initial dataset size: 18252 rows

Rows with greater than 6 missing values: 1

Final missing values:
Date            0
AveragePrice    0
Total Volume    0
4046            0
4225            0
4770            0
Total Bags      0
Small Bags      0
Large Bags      0
XLarge Bags     0
type            0
year            0
region          0
dtype: int64

Total missing: 0
Final dataset size: 18251 rows


2.4 Find the correlation between the variables and describe how the correlated values among the variables impact the model accuracy.


In [136]:
# Write your code here

import pandas as pd
import numpy as np

numeric_df = df.select_dtypes(include=[np.number]) # calculate correlation matrix for numeric columns
correlation_matrix = numeric_df.corr()

print("Correlation Matrix:") # display correlation matrix
print(correlation_matrix.round(3))

Correlation Matrix:
              AveragePrice  Total Volume   4046   4225   4770  Total Bags  \
AveragePrice         1.000        -0.193 -0.208 -0.173 -0.179      -0.177   
Total Volume        -0.193         1.000  0.978  0.974  0.872       0.963   
4046                -0.208         0.978  1.000  0.926  0.833       0.920   
4225                -0.173         0.974  0.926  1.000  0.888       0.906   
4770                -0.179         0.872  0.833  0.888  1.000       0.792   
Total Bags          -0.177         0.963  0.920  0.906  0.792       1.000   
Small Bags          -0.175         0.967  0.925  0.916  0.803       0.994   
Large Bags          -0.173         0.881  0.839  0.810  0.698       0.943   
XLarge Bags         -0.118         0.747  0.699  0.689  0.680       0.804   

              Small Bags  Large Bags  XLarge Bags  
AveragePrice      -0.175      -0.173       -0.118  
Total Volume       0.967       0.881        0.747  
4046               0.925       0.839        0.699  
4

Correlation measures the linear relationship between two variables. If two variables have high correlation, it makes it hard to tell the influence one has on the other. Models may misattribute the effects of one variable onto another.

#3. Exploratory Data Analysis (EDA)


3.1 Describe the variables
- Describe all variables in the dataset.
- For continuous variables: report **range (min, max), mean, median, and distribution**.
- For categorical variables: list unique values.

In [137]:
# Write your code here

import pandas as pd
import numpy as np

numeric_vars = df.select_dtypes(include=[np.number]).columns.tolist() # separate numeric variables
categorical_vars = df.select_dtypes(include=['object']).columns.tolist() # separate categorical variables

print("\nNumeric Variables:")

for var in numeric_vars:
    print(f"\n{var}:") # list numerical variables
    print(f"  Range:       [{df[var].min():.2f}, {df[var].max():.2f}]") # calculate range, mean, and median for each variable
    print(f"  Mean:        {df[var].mean():.2f}")
    print(f"  Median:      {df[var].median():.2f}")

    skewness = df[var].skew() # calculates skewness of variable
    if abs(skewness) < 0.5: # checks if skewness is close to 0
        dist = "approximately normal"
    elif skewness > 0.5:
        dist = "right-skewed (most values on left, long tail on right)"
    else:
        dist = "left-skewed (most values on right, long tail on left)"

    print(f"  Distribution: {dist}")

print("\nCategorical Variables:")

for var in categorical_vars: # list categorical variables
    print(f"\n{var}:")

    unique_values = df[var].unique() # find unique categorical values
    print(f"  Number of unique values: {len(unique_values)}")

    unique_list = [str(x) for x in unique_values if pd.notna(x)] # list unique categorical values
    print(f"  Unique values: {sorted(unique_list)}")


Numeric Variables:

AveragePrice:
  Range:       [0.44, 3.25]
  Mean:        1.41
  Median:      1.37
  Distribution: right-skewed (most values on left, long tail on right)

Total Volume:
  Range:       [84.56, 62505646.52]
  Mean:        850552.31
  Median:      107354.25
  Distribution: right-skewed (most values on left, long tail on right)

4046:
  Range:       [0.00, 22743616.17]
  Mean:        292983.95
  Median:      8645.30
  Distribution: right-skewed (most values on left, long tail on right)

4225:
  Range:       [0.00, 20470572.61]
  Mean:        295122.55
  Median:      29056.73
  Distribution: right-skewed (most values on left, long tail on right)

4770:
  Range:       [0.00, 2546439.11]
  Mean:        22837.27
  Median:      184.99
  Distribution: right-skewed (most values on left, long tail on right)

Total Bags:
  Range:       [0.00, 19373134.37]
  Mean:        239613.96
  Median:      39738.53
  Distribution: right-skewed (most values on left, long tail on right)

Smal

3.2 Inspect the earliest recorded date
- Find the earliest `Date`.
- Check if there are avocado prices recorded from the earliest date up to 2010.
- Comment: does the earliest data point look reasonable? Keep or remove?

In [138]:
# Write your code here

import pandas as pd
import numpy as np

df['Date'] = pd.to_datetime(df['Date']) # convert date to datetime to use min() (sorts chronologically)

earliest_date = df['Date'].min() # find earliest date in "Date"

print(f"Earliest date: {earliest_date}")

before_2011 = df[df['Date'] <= pd.Timestamp('2010-12-31')] # check for dates before 2011
print(f"\nRecords up to 2010: {len(before_2011)}")

earliest_records = df[df['Date'] == earliest_date] # show earliest records
print(f"\nEarliest record(s):")
print(earliest_records[['Date', 'AveragePrice', 'Total Volume', 'type', 'region']])

Earliest date: 1904-01-21 00:00:00

Records up to 2010: 1

Earliest record(s):
            Date  AveragePrice  Total Volume     type            region
18250 1904-01-21          1.87      13766.76  organic  WestTexNewMexico


This data point looks unreasonable. The date is so far from other entries in the dataset and has little relevance here. We should remove it as it would not have a major effect on data and would likely improve data quality.

3.3 Highest average price
- Find the highest value in "AveragePrice".
- Report which region it belongs to.
- Describe how you obtained the result.

In [139]:
# Write your code here

import pandas as pd
import numpy as np

max_price_idx = df['AveragePrice'].idxmax() # finds the index for maxium entry in "AveragePrice"
max_price_row = df.loc[max_price_idx] # finds the row the maximum average price belongs to

print(f"Highest Average Price: ${max_price_row['AveragePrice']:.2f}") # prints the maximum
print(f"Region: {max_price_row['region']}") # prints the region

Highest Average Price: $3.25
Region: SanFrancisco


I used idxmax() to get the index of the entry of with the maximum average price. Then, I get the row that contains the maximum average price using the index and .loc(). Finally, I print the "AveragePrice" entry and the "region" entry from this row.

3.4 Highest total volume
- Find the highest total volume of avocados.
- Report which region it belongs to.
- Describe how you obtained the result.

In [140]:
# Write your code here

import pandas as pd
import numpy as np

max_volume_idx = df['Total Volume'].idxmax() # finds the index for maxium entry in "Total Volume"
max_volume_row = df.loc[max_volume_idx] # finds the row the maximum total volume belongs to

print(f"Highest Total Volume: {max_volume_row['Total Volume']:.2f}") # prints the maximum
print(f"Region: {max_volume_row['region']}") # prints the region

Highest Total Volume: 62505646.52
Region: TotalUS


I used idxmax() to get the index of the entry of with the maximum volume. Then, I get the row that contains the maximum volume using the index and .loc(). Finally, I print the "Total Volume" entry and the "region" entry from this row.